File: 3425_고스택.cpp
#include <iostream>
#include <stack>

using namespace std;

#define MAX 1000000000

typedef long long ll;

string command[10000];
int numStorage[10000];
stack<ll> numStack;
string c;
int ci, ni, num;
bool isError;

void calculate() {
    ni = 0;
    isError = false;
    for (int j = 0; j < ci; j++) {
        if (command[j] == "NUM") {
            numStack.push(numStorage[ni++]);
        } else if (numStack.empty()) {
            isError = true;
            break;
        } else if (command[j] == "POP") {
            numStack.pop();
        } else if (command[j] == "INV") {
            ll num = numStack.top();
            numStack.pop();
            numStack.push(num * -1);
        } else if (command[j] == "DUP") {
            numStack.push(numStack.top());
        } else if (numStack.size() < 2) {
            isError = true;
            break;
        } else if (command[j] == "SWP") {
            ll n1 = numStack.top();
            numStack.pop();
            ll n2 = numStack.top();
            numStack.top() = n1;
            numStack.push(n2);
        } else if (command[j] == "ADD") {
            ll n1 = numStack.top();
            numStack.pop();
            numStack.top() += n1;
        } else if (command[j] == "SUB") {
            ll n1 = numStack.top();
            numStack.pop();
            numStack.top() -= n1;
        } else if (command[j] == "MUL") {
            ll n1 = numStack.top();
            numStack.pop();
            numStack.top() *= n1;
        } else if (command[j] == "DIV") {
            ll n1 = numStack.top();
            if (n1 == 0) {
                isError = true;
                break;
            }
            numStack.pop();
            numStack.top() /= n1;
        } else if (command[j] == "MOD") {
            ll n1 = numStack.top();
            if (n1 == 0) {
                isError = true;
                break;
            }
            numStack.pop();
            numStack.top() %= n1;
        }
    }
}

int main() {
    while(true) {
        // 케이스마다 명령어 입력
        ci = 0;
        ni = 0;
        while(true) {
            cin >> c;
            if (c == "QUIT") exit(0);
            else if (c == "END") break;
            else if (c == "NUM") {
                cin >> numStorage[ni++];
            }
            command[ci++] = c;
        }

        int cnt;
        cin >> cnt;
        
        // 숫자 입력 & 각 숫자마다 명령 목록 수행
        for (int i = 0; i < cnt; i++) {
            cin >> num;
            numStack.push(num);
            
            // 명령 목록 수행
            calculate();
            
            // 수행 후 결과 출력
            if (isError || numStack.size() != 1 || numStack.top() > MAX || numStack.top() < MAX * -1)
                cout << "ERROR" << endl;
            else cout << numStack.top() << endl;
            
            // stack 초기화
            while (!numStack.empty()) numStack.pop();
        }
        cout << endl;
    }
}

File: 3425_고스택.swift
// swift 시간복잡도
// append() - 평균: O(1), 최악(메모리 재할당): O(N)
// popLast() - O(1) // return optional
// removeLast() - O(1) // not optional
// removeAll() - O(N)
// swapAt(_:_:) - O(1)
// sort(), sorted() - O(NlogN): MergeSort와 InsertionSort를 기반으로 만든 TimeSort 사용
import Foundation

let MAX = 1000000000
var numStack = [CLongLong]()
var command = [String](repeating: "", count: 100000)
var numStorage = [Int](repeating: 0, count: 100000)

var ci = 0, ni = 0
var isError = false

while true {
    ci = 0;
    ni = 0;
    while true {
        let c = readLine()!
        if c == "QUIT" { exit(0) }
        else if c == "END" { break }
        else if c.contains("NUM") {
            let num = c.components(separatedBy: " ")[1]
            numStorage[ni] = Int(num)!
            ni += 1
        }
        command[ci] = c
        ci += 1
    }
    
    let cnt = Int(readLine()!)!
    for _ in 0..<cnt {
        let num = CLongLong(readLine()!)!
        numStack.append(num)
        
        calculate()
        
        if isError || numStack.count != 1 || numStack.last! > MAX || numStack.last! < MAX * -1 {
            print("ERROR")
        } else {
            print(numStack.last!)
        }
        
        numStack.removeAll()
    }
    print()
}

func calculate() {
    ni = 0
    isError = false
    for i in 0..<ci {
        if command[i].contains("NUM") {
            numStack.append(CLongLong(numStorage[ni]))
            ni += 1
        } else if numStack.isEmpty {
            isError = true
            break
        } else if command[i] == "POP" {
            numStack.removeLast()
        } else if command[i] == "INV" {
            numStack.append(numStack.removeLast() * -1)
        } else if command[i] == "DUP" {
            numStack.append(numStack.last!)
        } else if command.count < 2{
            isError = true
            break
        } else if command[i] == "SWP" {
            numStack.swapAt(numStack.count - 2, numStack.count - 1)
        } else if command[i] == "ADD" {
            let n = numStack.removeLast()
            numStack.append(numStack.removeLast() + n)
        } else if command[i] == "SUB" {
            let n = numStack.removeLast()
            numStack.append(numStack.removeLast() - n)
        } else if command[i] == "MUL" {
            let n = numStack.removeLast()
            numStack.append(numStack.removeLast() * n)
        } else if command[i] == "DIV" {
            let n = numStack.removeLast()
            if n == 0 {
                isError = true
                break
            }
            numStack.append(numStack.removeLast() / n)
        } else if command[i] == "MOD" {
            let n = numStack.removeLast()
            if n == 0 {
                isError = true
                break
            }
            numStack.append(numStack.removeLast() % n)
        }
    }
}
